File: ccms/src/data/block/cms_urcb.c
#include "data/block/cms_urcb.h"
#include "data/string/cms_visible_string.h"
#include "data/string/cms_octet_string.h"
#include "per/cms_sequence.h"

int cms_urcb_encode_stream(per_stream_t *s, const void *ptr) {
    const cms_urcb_t *pdu = (const cms_urcb_t*)ptr;
    int err;

    /* 0. OPTIONAL bitmap (1 field: owner) — X.691 §22 */
    bool opt_present[1] = {
        pdu->owner_present && pdu->owner_present->value && pdu->owner
    };
    err = (int)per_encode_optional_bitmap(s, opt_present, 1);
    if (err) return err;

    /* 1. rptID — VisibleString (SIZE(129)) */
    if (!pdu->rptID) return CMS_ERR;
    err = cms_visible_string_encode_stream(s, pdu->rptID, CMS_URCB_RPT_ID_MAX_LEN);
    if (err) return err;

    /* 2. rptEna — BOOLEAN */
    if (!pdu->rptEna) return CMS_ERR;
    err = cms_boolean_encode_stream(s, pdu->rptEna);
    if (err) return err;

    /* 3. datSet — ObjectReference */
    if (!pdu->datSet) return CMS_ERR;
    err = cms_object_reference_encode_stream(s, pdu->datSet);
    if (err) return err;

    /* 4. confRev — INT32U */
    if (!pdu->confRev) return CMS_ERR;
    err = cms_int32u_encode_stream(s, pdu->confRev);
    if (err) return err;

    /* 5. optFlds — RCBOptFlds */
    if (!pdu->optFlds) return CMS_ERR;
    err = cms_rcb_opt_flds_encode_stream(s, pdu->optFlds);
    if (err) return err;

    /* 6. bufTm — INT32U */
    if (!pdu->bufTm) return CMS_ERR;
    err = cms_int32u_encode_stream(s, pdu->bufTm);
    if (err) return err;

    /* 7. sqNum — INT16U */
    if (!pdu->sqNum) return CMS_ERR;
    err = cms_int16u_encode_stream(s, pdu->sqNum);
    if (err) return err;

    /* 8. trgOps — TriggerConditions */
    if (!pdu->trgOps) return CMS_ERR;
    err = cms_trigger_conditions_encode_stream(s, pdu->trgOps);
    if (err) return err;

    /* 9. intgPd — INT32U */
    if (!pdu->intgPd) return CMS_ERR;
    err = cms_int32u_encode_stream(s, pdu->intgPd);
    if (err) return err;

    /* 10. gi — BOOLEAN */
    if (!pdu->gi) return CMS_ERR;
    err = cms_boolean_encode_stream(s, pdu->gi);
    if (err) return err;

    /* 11. resv — BOOLEAN [14] */
    if (!pdu->resv) return CMS_ERR;
    err = cms_boolean_encode_stream(s, pdu->resv);
    if (err) return err;

    /* 12. owner — OCTET STRING (SIZE(0..64)) OPTIONAL (bitmap[0]) */
    if (opt_present[0]) {
        err = cms_octet_string_encode_stream(s, pdu->owner, CMS_URCB_OWNER_MAX_LEN);
        if (err) return err;
    }

    return CMS_OK;
}

int cms_urcb_decode_stream(per_stream_t *s, void *ptr) {
    cms_urcb_t *pdu = (cms_urcb_t*)ptr;
    int err;

    /* 0. OPTIONAL bitmap (1 field: owner) — X.691 §22 */
    bool opt_present[1] = {false};
    err = (int)per_decode_optional_bitmap(s, opt_present, 1);
    if (err) return err;
    if (pdu && pdu->owner_present)
        pdu->owner_present->value = opt_present[0] ? 1 : 0;

    /* 1. rptID */
    if (pdu && !pdu->rptID) return CMS_ERR;
    err = cms_visible_string_decode_stream(s, pdu ? pdu->rptID : NULL, CMS_URCB_RPT_ID_MAX_LEN);
    if (err) return err;

    /* 2. rptEna */
    if (pdu && !pdu->rptEna) return CMS_ERR;
    err = cms_boolean_decode_stream(s, pdu ? pdu->rptEna : NULL);
    if (err) return err;

    /* 3. datSet */
    if (pdu && !pdu->datSet) return CMS_ERR;
    err = cms_object_reference_decode_stream(s, pdu ? pdu->datSet : NULL);
    if (err) return err;

    /* 4. confRev */
    if (pdu && !pdu->confRev) return CMS_ERR;
    err = cms_int32u_decode_stream(s, pdu ? pdu->confRev : NULL);
    if (err) return err;

    /* 5. optFlds */
    if (pdu && !pdu->optFlds) return CMS_ERR;
    err = cms_rcb_opt_flds_decode_stream(s, pdu ? pdu->optFlds : NULL);
    if (err) return err;

    /* 6. bufTm */
    if (pdu && !pdu->bufTm) return CMS_ERR;
    err = cms_int32u_decode_stream(s, pdu ? pdu->bufTm : NULL);
    if (err) return err;

    /* 7. sqNum */
    if (pdu && !pdu->sqNum) return CMS_ERR;
    err = cms_int16u_decode_stream(s, pdu ? pdu->sqNum : NULL);
    if (err) return err;

    /* 8. trgOps */
    if (pdu && !pdu->trgOps) return CMS_ERR;
    err = cms_trigger_conditions_decode_stream(s, pdu ? pdu->trgOps : NULL);
    if (err) return err;

    /* 9. intgPd */
    if (pdu && !pdu->intgPd) return CMS_ERR;
    err = cms_int32u_decode_stream(s, pdu ? pdu->intgPd : NULL);
    if (err) return err;

    /* 10. gi */
    if (pdu && !pdu->gi) return CMS_ERR;
    err = cms_boolean_decode_stream(s, pdu ? pdu->gi : NULL);
    if (err) return err;

    /* 11. resv */
    if (pdu && !pdu->resv) return CMS_ERR;
    err = cms_boolean_decode_stream(s, pdu ? pdu->resv : NULL);
    if (err) return err;

    /* 12. owner — OCTET STRING OPTIONAL (bitmap[0]) */
    if (opt_present[0]) {
        err = cms_octet_string_decode_stream(s, pdu ? pdu->owner : NULL, CMS_URCB_OWNER_MAX_LEN);
        if (err) return err;
    }

    return CMS_OK;
}

int cms_urcb_encode(const void *ptr, uint8_t **out_buf, size_t *out_len) {
    per_stream_t s;
    per_error_t err = per_stream_init_write(&s, 64);
    if (err) return (int)err;
    int rc = cms_urcb_encode_stream(&s, ptr);
    if (rc) { per_stream_free(&s); return rc; }
    *out_buf = per_stream_detach(&s, out_len);
    return CMS_OK;
}

int cms_urcb_decode(void *ptr, const uint8_t *in_buf, int in_len) {
    per_stream_t s;
    per_stream_init_read(&s, in_buf, (size_t)in_len);
    return cms_urcb_decode_stream(&s, ptr);
}

File: jcms/jcms-core/src/main/java/com/ysh/jcms/core/data/sequence/block/CmsUrcb.java
package com.ysh.jcms.core.data.sequence.block;

import com.ysh.jcms.data.bitarray.CmsRcbOptFlds;
import com.ysh.jcms.data.bitarray.CmsTriggerConditions;
import com.ysh.jcms.data.core.CmsField;
import com.ysh.jcms.data.core.CmsSequence;
import com.ysh.jcms.data.*;
import com.ysh.jcms.data.scalar.*;
import com.ysh.jcms.data.scalar.CmsObjectReference;
import com.ysh.jcms.data.scalar.CmsOctetString;
import com.ysh.jcms.data.scalar.CmsString;

/**
 * <pre>
 * {@code
 * URCB ::= SEQUENCE {
 *     rptID           [1] IMPLICIT VisibleString129,
 *     rptEna          [2] IMPLICIT BOOLEAN,
 *     datSet          [3] IMPLICIT ObjectReference,
 *     confRev         [4] IMPLICIT INT32U,
 *     optFlds         [5] IMPLICIT RCBOptFlds,
 *     bufTm           [6] IMPLICIT INT32U,
 *     sqNum           [7] IMPLICIT INT16U,
 *     trgOps          [8] IMPLICIT TriggerConditions,
 *     intgPd          [9] IMPLICIT INT32U,
 *     gi              [10] IMPLICIT BOOLEAN,
 *     resv            [14] IMPLICIT BOOLEAN,
 *     owner           [15] IMPLICIT OCTET STRING (SIZE (0..64)) OPTIONAL
 * } — 8.7.4
 * }
 * </pre>
 */
public class CmsUrcb extends CmsSequence {
    @CmsField
    public CmsString rptID;
    @CmsField
    public CmsBoolean rptEna;
    @CmsField
    public CmsObjectReference datSet;
    @CmsField
    public CmsInt32U confRev;
    @CmsField
    public CmsRcbOptFlds optFlds;
    @CmsField
    public CmsInt32U bufTm;
    @CmsField
    public CmsInt16U sqNum;
    @CmsField
    public CmsTriggerConditions trgOps;
    @CmsField
    public CmsInt32U intgPd;
    @CmsField
    public CmsBoolean gi;
    @CmsField
    public CmsBoolean resv;
    @CmsField(optional = true)
    public CmsOctetString owner;

    public CmsUrcb() {
        super(new InnerURCB());
    }

    public CmsUrcb rptID(String v) {
        this.rptID.value(v);
        return this;
    }
    public CmsUrcb rptEna(boolean v) {
        this.rptEna.value(v);
        return this;
    }
    public CmsUrcb datSet(String v) {
        this.datSet.value(v);
        return this;
    }
    public CmsUrcb confRev(long v) {
        this.confRev.value(v);
        return this;
    }
    public CmsUrcb optFlds(CmsRcbOptFlds v) {
        this.optFlds.value(v);
        return this;
    }
    public CmsUrcb bufTm(long v) {
        this.bufTm.value(v);
        return this;
    }
    public CmsUrcb sqNum(int v) {
        this.sqNum.value(v);
        return this;
    }
    public CmsUrcb trgOps(CmsTriggerConditions v) {
        this.trgOps.value(v);
        return this;
    }
    public CmsUrcb intgPd(long v) {
        this.intgPd.value(v);
        return this;
    }
    public CmsUrcb gi(boolean v) {
        this.gi.value(v);
        return this;
    }
    public CmsUrcb resv(boolean v) {
        this.resv.value(v);
        return this;
    }
    public CmsUrcb owner(byte[] v) {
        if (v != null) {
            this.owner.value(v);
            setPresent("owner", true);
        } else {
            setPresent("owner", false);
        }
        return this;
    }

    /** Copy all field values from another CmsUrcb (fluent). */
    public CmsUrcb value(CmsUrcb v) {
        rptID(v.rptID.value());
        rptEna(v.rptEna.value());
        datSet(v.datSet.value());
        confRev(v.confRev.value());
        optFlds(v.optFlds);
        bufTm(v.bufTm.value());
        sqNum(v.sqNum.value());
        trgOps(v.trgOps);
        intgPd(v.intgPd.value());
        gi(v.gi.value());
        resv(v.resv.value());
        if (v.isPresent("owner")) {
            this.owner.value(v.owner.value());
            setPresent("owner", true);
        } else {
            setPresent("owner", false);
        }
        return this;
    }
}
